File: sdk/python/src/makra/_sse.py
"""A line-oriented Server-Sent Events decoder.

The decoder is deliberately transport-agnostic: it is fed decoded text lines
and emits complete events, so the synchronous and asynchronous clients can
share one implementation of the framing rules from the WHATWG SSE spec.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional


@dataclass(frozen=True)
class ServerSentEvent:
    """One dispatched SSE message."""

    event: str
    data: str
    id: Optional[str] = None
    retry: Optional[int] = None


class SSEDecoder:
    """Accumulates SSE field lines and dispatches them on a blank line."""

    def __init__(self) -> None:
        self._event = ""
        self._data: list[str] = []
        self._id: Optional[str] = None
        self._retry: Optional[int] = None
        self.last_event_id: Optional[str] = None

    def feed(self, line: str) -> Optional[ServerSentEvent]:
        """Consume one line and return an event when the message is complete."""
        line = line.rstrip("\r").lstrip("﻿")
        if not line:
            return self._dispatch()
        # A line beginning with a colon is a comment. The gateway sends
        # ": keepalive" every 15 seconds to hold the connection open.
        if line.startswith(":"):
            return None
        field, _, value = line.partition(":")
        if value.startswith(" "):
            value = value[1:]
        if field == "event":
            self._event = value
        elif field == "data":
            self._data.append(value)
        elif field == "id" and "\x00" not in value:
            self._id = value
            self.last_event_id = value
        elif field == "retry":
            try:
                self._retry = int(value)
            except ValueError:
                pass
        return None

    def _dispatch(self) -> Optional[ServerSentEvent]:
        if not self._data and not self._event:
            self._reset()
            return None
        event = ServerSentEvent(
            event=self._event or "message",
            data="\n".join(self._data),
            id=self._id,
            retry=self._retry,
        )
        self._reset()
        return event

    def _reset(self) -> None:
        self._event = ""
        self._data = []
        self._id = None

File: sdk/javascript/src/sse.js
/**
 * A line-oriented Server-Sent Events decoder.
 *
 * The decoder is deliberately transport-agnostic: it is fed decoded text lines
 * and emits complete events, so the framing rules from the WHATWG SSE spec
 * live in exactly one place.
 */

/** Accumulates SSE field lines and dispatches them on a blank line. */
export class SSEDecoder {
  #event = "";
  #data = [];
  #id;
  #retry;

  constructor() {
    this.lastEventId = undefined;
  }

  /** Consume one line and return an event when the message is complete. */
  feed(rawLine) {
    const line = rawLine.replace(/\r$/, "").replace(/^﻿/, "");
    if (!line) return this.#dispatch();
    // A line beginning with a colon is a comment. The gateway sends
    // ": keepalive" every 15 seconds to hold the connection open.
    if (line.startsWith(":")) return null;

    const separator = line.indexOf(":");
    const field = separator === -1 ? line : line.slice(0, separator);
    let value = separator === -1 ? "" : line.slice(separator + 1);
    if (value.startsWith(" ")) value = value.slice(1);

    if (field === "event") {
      this.#event = value;
    } else if (field === "data") {
      this.#data.push(value);
    } else if (field === "id" && !value.includes("\0")) {
      this.#id = value;
      this.lastEventId = value;
    } else if (field === "retry") {
      const retry = Number(value);
      if (Number.isInteger(retry)) this.#retry = retry;
    }
    return null;
  }

  #dispatch() {
    if (this.#data.length === 0 && !this.#event) {
      this.#reset();
      return null;
    }
    const event = {
      event: this.#event || "message",
      data: this.#data.join("\n"),
      id: this.#id,
      retry: this.#retry,
    };
    this.#reset();
    return event;
  }

  #reset() {
    this.#event = "";
    this.#data = [];
    this.#id = undefined;
  }
}

/**
 * Split a `fetch` response body into text lines.
 *
 * `onChunk` fires for every network chunk, including heartbeat comments, which
 * is what lets the caller treat silence — rather than slowness — as a dead
 * connection.
 */
export async function* readLines(body, onChunk) {
  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      onChunk?.();
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";
      yield* lines;
    }
    buffer += decoder.decode();
    if (buffer) yield buffer;
  } finally {
    reader.cancel().catch(() => {});
  }
}
